### PR TITLE
Only allow known and safe file formats

### PR DESCRIPTION
**Pull request by cloudhary**
## Description

Mandiant flagged that you can upload and distribute malicious files. This PR mitigates the issue by limiting file uploads to the following file types on the frontend:

- PDF: `application/pdf`
- Word, Excel, Powerpoint: see [this](https://filext.com/faq/office_mime_types.html)
- Images: `image/gif`, `image/jpeg`, `image/bmp`, `image/png`
- Audio: see [this](https://www.iana.org/assignments/media-types/media-types.xhtml#audio)
- Video: see [this](https://www.iana.org/assignments/media-types/media-types.xhtml#video)
- Text: `text/csv`, `text/plain'

It does so via the `<input accept=''>` HTML element field and also through a frontend controller check via MIME types and file extensions. The reason why the latter was also implemented was because users on certain OS/browsers could still use the default "All files" filter on their file explorer due to browser/OS implementation decisions. They could also just drag-n-drop a file which does not pass the `accept` field check. See this [issue](https://stackoverflow.com/questions/4328947/limit-file-format-when-using-input-type-file/4329103#4329103) for more information.

## Decisions/Tradeoffs

I decided on a whitelist instead of a blacklist because it's more secure than trying to make sure to disable all file formats that are possibly risky.

## Screenshots

image

## How to test

You will only be able to upload whitelisted file formats. Files such as `.dll`, `.exe` and `.zip` should not be accepted.

## Type of change

- New feature (non-breaking change which adds functionality)

## Deploy Notes

You can optionally specify the `ALLOWED_FILE_TYPES` env var. However, it is not necessary since it will use the default specification for acceptable file types/extensions if the env var is not specified.

### CODE REVIEW

**Comment by DigiPie** [comment]
Some screenshots!

image

image

**Comment by cloudhary** [comment]
Erhm, what a failure on my part. Let me go back to the drawing board.

File: app/controller.js [hunk 1]
```diff
@@ -90,6 +90,42 @@ export default function(state, emitter) {
...
+        render();
+        return;
+      }
+
+      // Check for valid extension via MIME type or filename extension substring
+      let ft = files[i].type; // MIME type
+      let fe = fn.substring(extIndex); // Extension part of filename
+      let isInvalidType = true;
+      for (let j = 0; j < fileTypes.length; j++) {
+        if (ft === fileTypes[j] || fe === fileTypes[j]) {
```
Comment (DigiPie) [inline app/controller.js:114]: @cloudhary could you take a look at this and lmk what you think?

Comment (cloudhary) [reply app/controller.js:114]: I'm alright with it for now, we can wait to hear back from Mandiant to figure out if there's a more robust way forward

**Comment by DigiPie** [comment]
Ready for review! @cloudhary

File: server/config.js [hunk 2]
```diff
@@ -69,6 +69,269 @@ const conf = convict({
     default: 16,
     env: 'MAX_ARCHIVES_PER_USER'
   },
+  allowed_file_types: {
```
Comment (cloudhary) [inline server/config.js:72]: This is just too hilarious

**Comment by cloudhary** [comment]
Nice!

image

But for some reason it lets me click on a .md file
image

and eventually disallows it, but well, it's no big deal.

Since I was the "author" of this PR, even though you did all the work, could you tag @danielkhoo to review and approve this PR?

File: server/config.js (hunk 2 above)
Comment (DigiPie) [reply server/config.js:72]: 😆 don't judge!

**Comment by DigiPie** [comment]
Thanks @danielkhoo !

**Pull request merged**